include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -6321,12 +6326,10 @@ namespace VectorTools
                                   "with at least 'dim' components. In that case, this function "
                                   "will take the divergence of the first 'dim' components."));
               double sum = 0;
-              // take the trace of the derivatives, square it, multiply it
-              // with the weight function
+              // take the trace of the derivatives scaled by the weight and square it
               for (unsigned int k=0; k<dim; ++k)
-                sum += (data.psi_grads[q][k][k] * data.psi_grads[q][k][k]) *
-                       data.weight_vectors[q](k);
-              diff += sum * fe_values.JxW(q);
+                sum += data.psi_grads[q][k][k] * std::sqrt(data.weight_vectors[q](k));
+              diff += sum * sum * fe_values.JxW(q);
```

Comment:
I think I finally understand why you wrote it this way: when taking the trace (the divergence) you think that the weights scale the individual vector components. That clearly makes no sense. I think the only reasonable choice here is actually if the weight is scalar. In that case, all components of the `weight_vector` are the same and you can pull the square root out and just multiply by the weight in line 6332.